service/src/DCGMDevicePool.cpp

```
@@ -0,0 +1,160 @@
+        unsigned int dcgm_dev_id_list[DCGM_MAX_NUM_DEVICES];
+        result = dcgmGetAllSupportedDevices(m_dcgm_handle, dcgm_dev_id_list, &m_dcgm_dev_count);
+        check_dcgm_result(result, "Error fetching DCGM supported devices.");
+
+        for(int i = 0; i < m_dcgm_dev_count; ++i) {
+            std::vector<dcgmFieldValue_v1> field_values(sizeof(m_dcgm_field_ids)/sizeof(m_dcgm_field_ids[0]));
+            m_dcgm_field_values.push_back(field_values);;
+        }
+
+        //Setup Field Group
+        result = dcgmFieldGroupCreate(m_dcgm_handle, sizeof(m_dcgm_field_ids)/sizeof(m_dcgm_field_ids[0]), &m_dcgm_field_ids[0],
+                                      (char *)"geopm_fields", &m_field_group_id);
+        check_dcgm_result(result, "Error creating DCGM geopm_fields group.");
+
+        // Note: Currently we are using dcgmWatchFields, but may transition to using
+        //       dcgmProfWatchFields at a later date
+        result = dcgmWatchFields(m_dcgm_handle, DCGM_GROUP_ALL_GPUS, m_field_group_id, m_update_freq,
+                                 m_max_keep_age, m_max_keep_sample);
+
+        check_dcgm_result(result, "Error setting default watch field configuration.");
+    }
+
+    DCGMDevicePoolImp::~DCGMDevicePoolImp()
+    {
+        //Shutdown DCGM
+        dcgmStatusDestroy(NULL);
+        dcgmGroupDestroy(m_dcgm_handle, DCGM_GROUP_ALL_GPUS);
+    }
+
+    void DCGMDevicePoolImp::check_dcgm_result(const dcgmReturn_t result, const std::string error)
+    {
+        if (result != DCGM_ST_OK) {
+            throw Exception("DCGMDevicePool::" + std::string(__func__) + ": "
+                            + error + ": " + errorString(result),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+    }
+
+    int DCGMDevicePoolImp::dcgm_device() const {
+        return m_dcgm_dev_count;
+    }
+
+    double DCGMDevicePoolImp::field_value(int accel_idx, int geopm_field_id) const {
+        double result;
```

Maybe throw an exception in debug builds if type is not DCGM_FT_DOUBLE

---

Maybe also a not-just-debug check that all fields are double at initialization time, if that's possible.

---

Moved to doing prechecking of all fields at init time.  Removed the PCIE_* fields as they are NOT double formatted.  We can re-add them if needed and add additional checking for the appropriate data type.